### PR TITLE
AzCopyCore updates

### PR DESCRIPTION
**Pull request by KrzysztofCwalina**
- updated the sample to new packages and runtime
- refactored HTTP code into a separate wrapper for IDuplexPipe
- small resulting renames

### CODE REVIEW

File: samples/AzCopyCore/AzCopyCore/AzCopyCore.csproj [hunk 1]
```diff
@@ -19,18 +21,19 @@
...
   </ItemGroup>
 
   <ItemGroup>
-    <PackageReference Include="System.Azure.Experimental" Version="$(CoreFxLabPackageVersion)" />
-    <PackageReference Include="System.Buffers.Experimental" Version="$(CoreFxLabPackageVersion)" />
-    <PackageReference Include="System.Text.Http" Version="$(CoreFxLabPackageVersion)" />
-    <PackageReference Include="System.Text.Http.Parser" Version="$(CoreFxLabPackageVersion)" />
-    <PackageReference Include="System.Text.Utf8String" Version="$(CoreFxLabPackageVersion)" />
-    <PackageReference Include="System.IO.Pipelines" Version="$(CoreFxLabPackageVersion)" />
+    <PackageReference Include="System.Azure.Experimental" Version="0.1.0-preview2-180320-2" />
```
Comment (ahsonkhan) [inline samples/AzCopyCore/AzCopyCore/AzCopyCore.csproj]: Update and use the CoreFxLabPackageVersion to make it easier to update in the future.

File: samples/AzCopyCore/AzCopyCore/AzCopyCore.csproj [hunk 2]
```diff
@@ -19,18 +21,19 @@
   </ItemGroup>
 
   <ItemGroup>
-    <PackageReference Include="Microsoft.DotNet.ILCompiler" Version="1.0.0-alpha-26223-02" />
-    <PackageReference Include="System.Memory" Version="$(CoreFxPackageVersion)" />
-    <PackageReference Include="System.Threading.Tasks.Extensions" Version="$(CoreFxPackageVersion)" />
+    <PackageReference Include="Microsoft.DotNet.ILCompiler" Version="1.0.0-alpha-26320-02" />
+    <PackageReference Include="System.Buffers.ReaderWriter" Version="0.1.0-preview2-180320-2" />
```
Comment (ahsonkhan) [inline samples/AzCopyCore/AzCopyCore/AzCopyCore.csproj]: nit: Move this down to the other ItemGroup with the rest of the corefxlab packages

File: samples/AzCopyCore/AzCopyCore/AzCopyCore.csproj [hunk 3]
```diff
@@ -19,18 +21,19 @@
...
-    <PackageReference Include="System.Text.Http" Version="$(CoreFxLabPackageVersion)" />
-    <PackageReference Include="System.Text.Http.Parser" Version="$(CoreFxLabPackageVersion)" />
-    <PackageReference Include="System.Text.Utf8String" Version="$(CoreFxLabPackageVersion)" />
-    <PackageReference Include="System.IO.Pipelines" Version="$(CoreFxLabPackageVersion)" />
+    <PackageReference Include="System.Azure.Experimental" Version="0.1.0-preview2-180320-2" />
+    <PackageReference Include="System.Buffers.Experimental" Version="0.1.0-preview2-180320-2" />
+    <PackageReference Include="System.Text.Http" Version="0.1.0-preview2-180320-2" />
+    <PackageReference Include="System.Text.Http.Parser" Version="0.1.0-preview2-180320-2" />
+    <PackageReference Include="System.Text.Utf8String" Version="0.1.0-preview2-180320-2" />
+    <PackageReference Include="System.IO.Pipelines" Version="4.5.0-preview3-26319-04" />
```
Comment (ahsonkhan) [inline samples/AzCopyCore/AzCopyCore/AzCopyCore.csproj]: nit: Move this up to the other ItemGroup with the rest of the CoreFX packages

File: samples/AzCopyCore/AzCopyCore/AzCopyCore.csproj [hunk 4]
```diff
@@ -19,18 +21,19 @@
   </ItemGroup>
 
   <ItemGroup>
-    <PackageReference Include="Microsoft.DotNet.ILCompiler" Version="1.0.0-alpha-26223-02" />
-    <PackageReference Include="System.Memory" Version="$(CoreFxPackageVersion)" />
-    <PackageReference Include="System.Threading.Tasks.Extensions" Version="$(CoreFxPackageVersion)" />
+    <PackageReference Include="Microsoft.DotNet.ILCompiler" Version="1.0.0-alpha-26320-02" />
+    <PackageReference Include="System.Buffers.ReaderWriter" Version="0.1.0-preview2-180320-2" />
+    <PackageReference Include="System.Memory" Version="4.5.0-preview3-26319-04" />
+    <PackageReference Include="System.Threading.Tasks.Extensions" Version="4.5.0-preview3-26319-04" />
```
Comment (ahsonkhan) [inline samples/AzCopyCore/AzCopyCore/AzCopyCore.csproj]: Update and use the CoreFxPackageVersion property to make it easier to update in the future.

File: samples/AzCopyCore/AzCopyCore/AzCopyCore.csproj [hunk 5]
```diff
@@ -3,7 +3,9 @@
   <PropertyGroup>
     <OutputType>Exe</OutputType>
     <TargetFramework>netcoreapp2.1</TargetFramework>
+	<!--
```
Comment (ahsonkhan) [inline samples/AzCopyCore/AzCopyCore/AzCopyCore.csproj]: Why is this commented out? This should be the same as https://github.com/dotnet/corefxlab/blob/master/SharedRuntimeVersion.txt (or import the common.props file).

File: samples/AzCopyCore/AzCopyCore/AzCopyCore.csproj [hunk 6]
```diff
@@ -19,18 +21,19 @@
...
 
   <ItemGroup>
-    <PackageReference Include="System.Azure.Experimental" Version="$(CoreFxLabPackageVersion)" />
-    <PackageReference Include="System.Buffers.Experimental" Version="$(CoreFxLabPackageVersion)" />
-    <PackageReference Include="System.Text.Http" Version="$(CoreFxLabPackageVersion)" />
-    <PackageReference Include="System.Text.Http.Parser" Version="$(CoreFxLabPackageVersion)" />
-    <PackageReference Include="System.Text.Utf8String" Version="$(CoreFxLabPackageVersion)" />
-    <PackageReference Include="System.IO.Pipelines" Version="$(CoreFxLabPackageVersion)" />
+    <PackageReference Include="System.Azure.Experimental" Version="0.1.0-preview2-180320-2" />
+    <PackageReference Include="System.Buffers.Experimental" Version="0.1.0-preview2-180320-2" />
```
Comment (ahsonkhan) [inline samples/AzCopyCore/AzCopyCore/AzCopyCore.csproj]: Can these be project references instead of package references?

File: samples/AzCopyCore/AzCopyCore/Program.cs [hunk 7]
```diff
@@ -28,22 +28,19 @@ static void PrintUsage()
...
-
-        long before = GC.GetAllocatedBytesForCurrentThread();
+        Log.Switch.Level = SourceLevels.Information;
 
         var options = new CommandLine(args);
         ReadOnlySpan<char> source = options.GetSpan("/Source:");
         ReadOnlySpan<char> destination = options.GetSpan("/Dest:");
 
+        long before = GC.GetAllocatedBytesForCurrentThread();
+        var sw = Stopwatch.StartNew();
```
Comment (ahsonkhan) [inline samples/AzCopyCore/AzCopyCore/Program.cs]: nit: use of var

File: samples/AzCopyCore/AzCopyCore/Program.cs [hunk 8]
```diff
@@ -96,9 +101,9 @@ static void TransferDirectoryToStorage(ReadOnlySpan<char> localDirectory, ReadOn
         ReadOnlySpan<char> host = storageFullPath.Slice(0, pathStart);
         ReadOnlySpan<char> path = storageFullPath.Slice(pathStart + 1);
 
-        using (var client = new StorageClient(keyBytes, account, host))
+        using (var client = new StorageClient(keyBytes, account, host, 80, Log))
         {
-            client.Log = Log;
+            //var files = new Files(directoryPath);
```
Comment (ahsonkhan) [inline samples/AzCopyCore/AzCopyCore/Program.cs]: clean up commented out code

File: samples/AzCopyCore/AzCopyCore/Program.cs [hunk 9]
```diff
@@ -76,7 +79,9 @@ static void TransferDirectoryToStorage(ReadOnlySpan<char> localDirectory, ReadOn
         byte[] keyBytes;
         if (options.Contains("/DestKey:"))
         {
-            keyBytes = options["/DestKey:"].ComputeKeyBytes();
+            ReadOnlySpan<char> key = options["/DestKey:"];
+            var str = key.ToString();
+            keyBytes = key.ComputeKeyBytes();
```
Comment (ahsonkhan) [inline samples/AzCopyCore/AzCopyCore/Program.cs]: Why is this change necessary?

File: samples/AzCopyCore/AzCopyCore/AzCopyCore.csproj (hunk 1 above)
Comment (KrzysztofCwalina) [reply samples/AzCopyCore/AzCopyCore/AzCopyCore.csproj]: Yeah, but there is a trade off here. I think it's good that we have some samples that validate how our customers are using and demonstrate how  they should be using our packages.

File: samples/AzCopyCore/AzCopyCore/AzCopyCore.csproj (hunk 2 above)
Comment (KrzysztofCwalina) [reply samples/AzCopyCore/AzCopyCore/AzCopyCore.csproj]: will do

File: samples/AzCopyCore/AzCopyCore/AzCopyCore.csproj (hunk 3 above)
Comment (KrzysztofCwalina) [reply samples/AzCopyCore/AzCopyCore/AzCopyCore.csproj]: will do

File: samples/AzCopyCore/AzCopyCore/AzCopyCore.csproj (hunk 4 above)
Comment (KrzysztofCwalina) [reply samples/AzCopyCore/AzCopyCore/AzCopyCore.csproj]: As I said above, I like how the sample shows VS experience that our customers will get.

File: samples/AzCopyCore/AzCopyCore/AzCopyCore.csproj (hunk 5 above)
Comment (KrzysztofCwalina) [reply samples/AzCopyCore/AzCopyCore/AzCopyCore.csproj]: It's @terrajobst's fault. :-) I removed it now.

File: samples/AzCopyCore/AzCopyCore/AzCopyCore.csproj (hunk 6 above)
Comment (KrzysztofCwalina) [reply samples/AzCopyCore/AzCopyCore/AzCopyCore.csproj]: Same as above: I think samples should show customer experience that does not require cloning the whole repo.

File: samples/AzCopyCore/AzCopyCore/Program.cs (hunk 7 above)
Comment (KrzysztofCwalina) [reply samples/AzCopyCore/AzCopyCore/Program.cs]: will fix

File: samples/AzCopyCore/AzCopyCore/Program.cs (hunk 8 above)
Comment (KrzysztofCwalina) [reply samples/AzCopyCore/AzCopyCore/Program.cs]: will do.

File: samples/AzCopyCore/AzCopyCore/Program.cs (hunk 9 above)
Comment (KrzysztofCwalina) [reply samples/AzCopyCore/AzCopyCore/Program.cs]: leftover from debugging. I removed now. thanks.

File: samples/AzCopyCore/AzCopyCore/AzCopyCore.csproj (hunk 1 above)
Comment (ahsonkhan) [reply samples/AzCopyCore/AzCopyCore/AzCopyCore.csproj]: I don't understand the trade off here. We have all these package version numbers here, and they point to the same version. We have a property in *this* csproj called CoreFxLabPackageVersion. I was just suggesting we use that so that it becomes a single point of edit in the future, if we ever want to upgrade them again. Note, this is still isolated from the rest of the repo and would still demonstrate the customer's use of our packages.

File: samples/AzCopyCore/AzCopyCore/AzCopyCore.csproj (hunk 6 above)
Comment (ahsonkhan) [reply samples/AzCopyCore/AzCopyCore/AzCopyCore.csproj]: Agreed. In that case, package references make sense. Thanks.

File: samples/AzCopyCore/AzCopyCore/AzCopyCore.csproj (hunk 4 above)
Comment (ahsonkhan) [reply samples/AzCopyCore/AzCopyCore/AzCopyCore.csproj]: I don't know if someone would go and manually update each individual referenced packages using VS/NuGet package manager. Once they are added, it would make sense to create a property to update the versions, no?

File: samples/AzCopyCore/AzCopyCore/Program.cs (hunk 9 above)
Comment (ahsonkhan) [reply samples/AzCopyCore/AzCopyCore/Program.cs]: FYI, I just confirmed, debugging works as expected and shows the string. So, hopefully, these types of changes won't be necessary anymore.

image
